File: Function.cs

/*****************************************************************************
 *                                                                           *
 *  Function.cs                                                              *
 *  24 September 2006                                                        *
 *  Project: Solus, Ligra                                                    *
 *  Written by: Richard Sartor                                               *
 *  Copyright © 2006 Metaphysics Industries, Inc.                            *
 *                                                                           *
 *  Converted from C++ to C# on 29 October 2007                              *
 *                                                                           *
 *  A mathematical function that can be evaluated with a set of              *
 *    parameters. This serves as a base class that is inherited by other,    *
 *    specialized classes, each representing a different mathematical        *
 *    function (e.g. "SineFunction : Function"). This base class performs    *
 *    all necessary type checking on given arguments based on information    *
 *    specified by the derived class.                                        *
 *                                                                           *
 *****************************************************************************/

using System;
using System.Collections.Generic;
using MetaphysicsIndustries.Collections;
using MetaphysicsIndustries.Giza;

namespace MetaphysicsIndustries.Solus
{
    public abstract partial class Function
	{
        protected Function()
            : this(string.Empty)
        {
        }

        protected Function(string name)
        {
            _name = name;
        }

		public virtual Literal Call(SolusEnvironment env, params Expression[] args)
        {
            this.CheckArguments(args);
            List<Literal> evalArgs = new List<Literal>(args.Length);
            foreach (Expression arg in args)
            {
                evalArgs.Add(arg.Eval(env));
            }
			return this.InternalCall(env, evalArgs.ToArray());
		}

        //public virtual Expression CleanUp(Expression[] args)
        //{
        //    bool call = true;
        //    foreach (Expression arg in args)
        //    {
        //        if (!(arg is Literal))
        //        {
        //            call = false;
        //            break;
        //        }
        //    }

        //    if (call)
        //    {
        //        return Call(null, args);
        //    }

        //    return new FunctionCall(this, args);
        //}

		public virtual string DisplayName
		{
            get { return Name; }
		}

		public string Name
		{
			get
			{
				return _name;
			}
			protected set
			{
				if (_name != value)
				{
					_name = value;
				}
			}
		}

		public List<Type> Types
		{
			get
			{
				return InternalTypes;
			}
		}

        protected abstract Literal InternalCall(SolusEnvironment env, Literal[] args);

		protected virtual void CheckArguments(Expression[] args)
		{
			Expression[] _args = args;
			List<Type>	types;
			int				i;
			int				j;
			
			types = Types;
			if (args.Length != types.Count)
			{
				throw new InvalidOperationException("Wrong number of arguments given to " + DisplayName + " (given " + args.Length.ToString() + ", require " + Types.Count.ToString() + ")");
			}
			Type	e;
			e = typeof(Expression);
			j = args.Length;
			for (i = 0; i < j; i++)
			{
				if (!e.IsAssignableFrom(types[i]))
				{
					
					throw new InvalidOperationException("Required argument type " + i.ToString() + " is invalid (given \"" + types[i].Name + "\", require \"" + e.Name + ")");
				}
				if (!types[i].IsAssignableFrom(args[i].GetType()))
				{
					throw new InvalidOperationException("Argument " + ((i).ToString()) + " of wrong type (given \"" + args.GetType().Name + "\", require \"" + types[i].Name + ")");
				}
			}
		}

		protected List<Type> InternalTypes
		{
			get
			{
				return _internaltypes;
			}
		}

		private  List<Type> _internaltypes = new List<Type>();
		private  string     _name;

        public virtual string ToString(List<Expression> arguments)
        {
            Expression[] exprs = arguments.ToArray();

            string[] strs = Array.ConvertAll<Expression, string>(exprs, Expression.ToString);

            return DisplayName + "(" + string.Join(", ", strs) + ")";
        }

        public T As<T>()
            where T : Function
        {
            return this as T;
        }

        public virtual string DocString
        {
            get { return string.Empty; }
        }

        public virtual IEnumerable<Instruction> ConvertToInstructions(VariableToArgumentNumberMapper varmap, List<Expression> arguments)
        {
            throw new NotImplementedException();
        }
    }
}

File: Function.cpp

/*****************************************************************************
 *                                                                           *
 *  Function.cpp                                                             *
 *  24 September 2006                                                        *
 *  Project: Solus, Ligra                                                    *
 *  Written by: Richard Sartor                                               *
 *  Copyright © 2006 Metaphysics Industries                                  *
 *                                                                           *
 *  A mathematical function that can be evaluated with a set of              *
 *    parameters. This serves as a base class that is inherited by other,    *
 *    specialized classes, each representing a different mathematical        *
 *    function (e.g. SineFunction : Function). This base class performs      *
 *    all necessary type checking on given arguments based on information    *
 *    specified by the derived class.                                        *
 *                                                                           *
 *****************************************************************************/



#include "stdafx.h"
#include "Function.h"
#include "Literal.h"
#include "CosineFunction.h"
#include "SineFunction.h"
#include "TangentFunction.h"
#include "CotangentFunction.h"
#include "SecantFunction.h"
#include "CosecantFunction.h"
#include "ArccosineFunction.h"
#include "ArcsineFunction.h"
#include "ArctangentFunction.h"
#include "ArccotangentFunction.h"
#include "ArcsecantFunction.h"
#include "ArccosecantFunction.h"
#include "FloorFunction.h"
#include "CeilingFunction.h"




;
NAMESPACE_START
;



Function::Function(void)
{
	this->internaltypes = gcnew List<Type^>;
	this->internaltypes->Add(Expression::typeid);
}

Function::~Function(void)
{
	
}

Function^ Function::Cosine::get(void)
{
	return Function::cosine;
}
Function^ Function::Sine::get(void)
{
	return Function::sine;
}
Function^ Function::Tangent::get(void)
{
	return Function::tangent;
}
Function^ Function::Cotangent::get(void)
{
	return Function::cotangent;
}
Function^ Function::Secant::get(void)
{
	return Function::secant;
}
Function^ Function::Cosecant::get(void)
{
	return Function::cosecant;
}
Function^ Function::Arccosine::get(void)
{
	return Function::arccosine;
}
Function^ Function::Arcsine::get(void)
{
	return Function::arcsine;
}
Function^ Function::Arctangent::get(void)
{
	return Function::arctangent;
}
Function^ Function::Arccotangent::get(void)
{
	return Function::arccotangent;
}
Function^ Function::Arcsecant::get(void)
{
	return Function::arcsecant;
}
Function^ Function::Arccosecant::get(void)
{
	return Function::arccosecant;
}
Function^ Function::Floor::get(void)
{
	return Function::floor;
}
Function^ Function::Ceiling::get(void)
{
	return Function::ceiling;
}

static Function::Function(void)
{
	Function::cosine = gcnew CosineFunction;
	Function::sine = gcnew SineFunction;
	Function::tangent = gcnew TangentFunction;
	Function::cotangent = gcnew CotangentFunction;
	Function::secant = gcnew SecantFunction;
	Function::cosecant = gcnew CosecantFunction;
	Function::arccosine = gcnew ArccosineFunction;
	Function::arcsine = gcnew ArcsineFunction;
	Function::arctangent = gcnew ArctangentFunction;
	Function::arccotangent = gcnew ArccotangentFunction;
	Function::arcsecant = gcnew ArcsecantFunction;
	Function::arccosecant = gcnew ArccosecantFunction;
	Function::floor = gcnew FloorFunction;
	Function::ceiling = gcnew CeilingFunction;
}

Literal^ Function::Call( ... array<Expression^>^ args)
{
	this->CheckArguments(args);

	return this->InternalCall(args);
}

//Literal^ Function::Call(ICollection<Expression^>^ )
//{
//	return gcnew Literal(0);
//}

List<Type^>^ Function::Types::get(void)
{
	return this->InternalTypes;
}
List<Type^>^ Function::InternalTypes::get(void)
{
	return this->internaltypes;
}

String^ Function::Name::get(void)
{
	return this->name;
}
void Function::Name::set(String^ s)
{
	if (this->name != s)
	{
		this->name = s;
	}
}

String^ Function::DisplayName::get(void)
{
	return this->Name;
}

void Function::CheckArguments( ... array<Expression^>^ args)
{
	array<Expression^>^ _args = args;

	List<Type^>^	types;
	int				i;
	int				j;
	//Type^			t;

	types = this->Types;

	if (args->Length != types->Count)
	{
		throw gcnew InvalidOperationException("Wrong number of arguments given to " + this->DisplayName + " (given " + args->Length.ToString() + ", require " + this->Types->Count.ToString() + ") " + __WCODESIG__);
	}

	Type^	e;

	e = Expression::typeid;

	j = args->Length;
	for (i = 0; i < j; i++)
	{
		if (!e->IsAssignableFrom(types[i]))
		{
			//change this so it's only checked when the types are set by the Function's constructor
			throw gcnew InvalidOperationException("Required argument type " + i.ToString() + " is invalid (given \"" + types[i]->Name + "\", require \"" + e->Name + ") " + __WCODESIG__);
		}
		if (!types[i]->IsAssignableFrom(args[i]->GetType()))
		{
			throw gcnew InvalidOperationException("Argument " + ((i).ToString()) + " of wrong type (given \"" + args->GetType()->Name + "\", require \"" + types[i]->Name + ") " + __WCODESIG__);
		}
	}
}


;
NAMESPACE_END
;


